### PR TITLE
[WICKET-6876] add an AJAX behavior to transfer client side file infor…

### PR DESCRIPTION
**Pull request by reiern70**
[WICKET-6876] add an AJAX behavior to transfer client side file information to server side (when a file is selected for upload at client side).

This PR replaces 

https://github.com/apache/wicket/pull/466

I have included comments/remarks provided there. Now also submit buttons will be disabled is chosen file is bigger than allowed

image

and enabled in case file is Ok

image

Thus this illustrates how AJAX behavior can be used to do on the spot "minimum validation" without need to submit file.

### CODE REVIEW

File: wicket-core/src/main/java/org/apache/wicket/markup/html/form/upload/FileUploadField.java
Comment (martin-g) [inline wicket-core/src/main/java/org/apache/wicket/markup/html/form/upload/FileUploadField.java:102]: I think we should think about the `multiple` support now, because later this method signature may need to be changed to something like `protected abstract void onFilesSelected(AjaxRequestTarget target, List<FileDescription>)`

Comment (reiern70) [reply wicket-core/src/main/java/org/apache/wicket/markup/html/form/upload/FileUploadField.java:102]: Oh. I was thinking to roll a different behavior. But you are probably right. I will add support for multiple values ASAP

Comment (reiern70) [reply wicket-core/src/main/java/org/apache/wicket/markup/html/form/upload/FileUploadField.java]: My latest commit adds support for multiple files. I have tested it locally and it seems to work (with multiple files). Not sure if the way I did it is the best/correct one.

**Comment by svenmeier** [comment]
IMHO OnFileSelectedBehavior deserves to be a top-level class. Maybe with its own JS resource, since the script to collect dynamicExtraParameters is getting very long now. 

With a static factory method users could use a lambda expression (like we have for several other behaviors, e.g. AjaxFormSubmitBehavior#onSubmit):

    file.add(OnFileSelectedBehavior.onSelected((AjaxRequestTarget target, List<OnFileSelectedBehavior.FileDescription> fileDescriptions) -> {
        // ...
    }));
	
Maybe drop the "On"-prefix too, since almost all current behaviors don't have it - are you doing Android development? ;)

File: wicket-core/src/main/java/org/apache/wicket/markup/html/form/upload/FileUploadField.java
Comment (martin-g) [inline wicket-core/src/main/java/org/apache/wicket/markup/html/form/upload/FileUploadField.java:145]: Prefer `Long.parseLong(String s, int radix)` instead

File: wicket-core/src/main/java/org/apache/wicket/markup/html/form/upload/FileUploadField.java
Comment (martin-g) [inline wicket-core/src/main/java/org/apache/wicket/markup/html/form/upload/FileUploadField.java:163]: s/onFileSelected/onSelect/ ? I think @svenmeier also suggested this name

File: wicket-core/src/main/java/org/apache/wicket/markup/html/form/upload/FileUploadField.java
Comment (martin-g) [inline wicket-core/src/main/java/org/apache/wicket/markup/html/form/upload/FileUploadField.java:185]: missing `var` for `f`

File: wicket-core/src/main/java/org/apache/wicket/markup/html/form/upload/FileUploadField.java
Comment (martin-g) [inline wicket-core/src/main/java/org/apache/wicket/markup/html/form/upload/FileUploadField.java:197]: Debug leftover

File: wicket-core/src/main/java/org/apache/wicket/markup/html/form/upload/FileUploadField.java
Comment (martin-g) [inline wicket-core/src/main/java/org/apache/wicket/markup/html/form/upload/FileUploadField.java:173]: Let's put the new property to `window.Wicket` namespace instead of to the global `window` namespace.

**Review by martin-g — APPROVED** [review]

**Comment by reiern70** [comment]
@svenmeier and @martin-g 

Thanks for your comments:

> IMHO OnFileSelectedBehavior deserves to be a top-level class.

 My rationale was the following: this class will only be used in relation with FileUploadField, why then pollute package? Of course, this is a matter of taste and consistency with project practices/conventions. Moving it top-level class.

> Maybe with its own JS resource, since the script to collect dynamicExtraParameters is getting very long now.

True.
 
> With a static factory method users could use a lambda expression (like we have for several other behaviors, e.g. AjaxFormSubmitBehavior#onSubmit):
> 
> ```
> file.add(OnFileSelectedBehavior.onSelected((AjaxRequestTarget target, List<OnFileSelectedBehavior.FileDescription> fileDescriptions) -> {
>     // ...
> }));
> ```
> 
> Maybe drop the "On"-prefix too, since almost all current behaviors don't have it - are you doing Android development? ;)

Never did Android development in my life. If you both agree this should be done I will gladly do it ;-)

**Comment by reiern70** [comment]
**Note:** A just thought that a possible addition in case of multiple file uploads is a component with a table displaying all the selected files. WDYT?

**Comment by reiern70** [comment]
> IMHO OnFileSelectedBehavior deserves to be a top-level class. Maybe with its own JS resource, since the script to collect dynamicExtraParameters is getting very long now.
> 
> With a static factory method users could use a lambda expression (like we have for several other behaviors, e.g. AjaxFormSubmitBehavior#onSubmit):
> 
> ```
> file.add(OnFileSelectedBehavior.onSelected((AjaxRequestTarget target, List<OnFileSelectedBehavior.FileDescription> fileDescriptions) -> {
>     // ...
> }));
> ```
> 
> Maybe drop the "On"-prefix too, since almost all current behaviors don't have it - are you doing Android development? ;)

image

I see a lot of 'On' hanging around. Thus unless there is a very strong opinion about it from most of the developers I will keep it.

File: wicket-core/src/main/java/org/apache/wicket/markup/html/form/upload/FileUploadField.java
Comment (reiern70) [reply wicket-core/src/main/java/org/apache/wicket/markup/html/form/upload/FileUploadField.java:163]: ok.

File: wicket-core/src/main/java/org/apache/wicket/markup/html/form/upload/FileUploadField.java
Comment (reiern70) [reply wicket-core/src/main/java/org/apache/wicket/markup/html/form/upload/FileUploadField.java:173]: Also I though about cleaning garbage after use it: this only needs to be kept in context of this AJAX request. Maybe we can even use attrs object as a holder. I will check.

File: wicket-core/src/main/java/org/apache/wicket/markup/html/form/upload/FileUploadField.java
Comment (reiern70) [inline wicket-core/src/main/java/org/apache/wicket/markup/html/form/upload/FileUploadField.java:138]: To myself: can we use getParameterValues instead?

**Comment by reiern70** [comment]
@svenmeier and @martin-g 

Thanks for your feedback and useful comments. I have reworked PR taking into account your comments. Can you please review again?

**Comment by reiern70** [comment]
The new section on examples looks like 

image

**Comment by reiern70** [comment]
> @svenmeier and @martin-g
> 
> Thanks for your comments:
> 
> > IMHO OnFileSelectedBehavior deserves to be a top-level class.
> 
> My rationale was the following: this class will only be used in relation with FileUploadField, why then pollute package? Of course, this is a matter of taste and consistency with project practices/conventions. Moving it top-level class.
> 
> > Maybe with its own JS resource, since the script to collect dynamicExtraParameters is getting very long now.
> 
> True.
> 
> > With a static factory method users could use a lambda expression (like we have for several other behaviors, e.g. AjaxFormSubmitBehavior#onSubmit):
> > ```
> > file.add(OnFileSelectedBehavior.onSelected((AjaxRequestTarget target, List<OnFileSelectedBehavior.FileDescription> fileDescriptions) -> {
> >     // ...
> > }));
> > ```
> > 
> > 
> > Maybe drop the "On"-prefix too, since almost all current behaviors don't have it - are you doing Android development? ;)
> 
> Never did Android development in my life. If you both agree this should be done I will gladly do it ;-)

@svenmeier My latest commit introduces factory method as  what you suggested

image

thanks for pointing this out.

**Comment by reiern70** [comment]
My IDE suggest me to use let instead of var

image

I was using let, reverting to use var otherwise build fails.

**Comment by reiern70** [comment]
once reviewed and approved I will rebase on top of master and squash commits.

**Comment by reiern70** [comment]
rebased on top oof latest master + squashed commits. Waiting for ok to merge )or do more changes if asked)

**Comment by martin-g** [comment]
> **Note:** A just thought that a possible addition in case of multiple file uploads is a component with a table displaying all the selected files. WDYT?

I don't think we should add such component. Every application may want to display it in a different way.

File: wicket-core/src/main/java/org/apache/wicket/markup/html/form/upload/FileUploadField.java
Comment (martin-g) [inline wicket-core/src/main/java/org/apache/wicket/markup/html/form/upload/FileUploadField.java:38]: It seems these imports are no more used.

File: wicket-core/src/main/java/org/apache/wicket/markup/html/form/upload/OnFilesSelectedBehavior.java
Comment (martin-g) [inline wicket-core/src/main/java/org/apache/wicket/markup/html/form/upload/OnFilesSelectedBehavior.java:47]: this line should be removed

File: wicket-core/src/main/java/org/apache/wicket/markup/html/form/upload/OnFilesSelectedBehavior.java
Comment (martin-g) [inline wicket-core/src/main/java/org/apache/wicket/markup/html/form/upload/OnFilesSelectedBehavior.java:46]: ... selected file`(s)` before uploading it/them.

File: wicket-core/src/main/java/org/apache/wicket/markup/html/form/upload/OnFilesSelectedBehavior.java
Comment (martin-g) [inline wicket-core/src/main/java/org/apache/wicket/markup/html/form/upload/OnFilesSelectedBehavior.java:108]: Any reason `super.**` to be called after ? I don't think there is any difference which one is first in this case. It just looks suspicious this way.

File: wicket-core/src/main/java/org/apache/wicket/markup/html/form/upload/OnFilesSelectedBehavior.java
Comment (martin-g) [inline wicket-core/src/main/java/org/apache/wicket/markup/html/form/upload/OnFilesSelectedBehavior.java:120]: For consistency with the other similar methods the method name should be `onSelect`
https://github.com/apache/wicket/blob/048e5681df960018722237ee6254273e839b5fd2/wicket-core/src/main/java/org/apache/wicket/ajax/AjaxEventBehavior.java#L166

File: wicket-core/src/main/java/org/apache/wicket/markup/html/form/upload/OnFilesSelectedBehavior.js
Comment (martin-g) [inline wicket-core/src/main/java/org/apache/wicket/markup/html/form/upload/OnFilesSelectedBehavior.js:37]: s/oor/or/

File: wicket-core/src/main/java/org/apache/wicket/markup/html/form/upload/OnFilesSelectedBehavior.js
Comment (martin-g) [inline wicket-core/src/main/java/org/apache/wicket/markup/html/form/upload/OnFilesSelectedBehavior.js:39]: what is the type of `component` here ? 
This parameter name is a little confusing. It sounds like a Wicket Component, but it is something else.

File: wicket-core/src/main/java/org/apache/wicket/markup/html/form/upload/OnFilesSelectedBehavior.js
Comment (martin-g) [inline wicket-core/src/main/java/org/apache/wicket/markup/html/form/upload/OnFilesSelectedBehavior.js:46]: Can we use `Wicket.OnFilesSelected[id]` instead of `window[id]` ?
We should try to not pollute the global namespace, even temporarily.

File: wicket-core/src/main/java/org/apache/wicket/markup/html/form/upload/OnFilesSelectedBehavior.js
Comment (martin-g) [inline wicket-core/src/main/java/org/apache/wicket/markup/html/form/upload/OnFilesSelectedBehavior.js:53]: s/IDs/details/ ?

File: wicket-examples/src/main/java/org/apache/wicket/examples/ajax/builtin/FileUploadPage$MultipleFileUploadsSamplePanel.html
Comment (martin-g) [inline wicket-examples/src/main/java/org/apache/wicket/examples/ajax/builtin/FileUploadPage$MultipleFileUploadsSamplePanel.html:16]: dangling `s` at the end of the line

File: wicket-examples/src/main/java/org/apache/wicket/examples/ajax/builtin/EffectsPage.html
Comment (martin-g) [inline wicket-examples/src/main/java/org/apache/wicket/examples/ajax/builtin/EffectsPage.html:3]: Why EffectsPage.html is modified ?

**Review by martin-g — CHANGES_REQUESTED** [review]

File: wicket-core/src/main/java/org/apache/wicket/markup/html/form/upload/OnFilesSelectedBehavior.java
Comment (reiern70) [reply wicket-core/src/main/java/org/apache/wicket/markup/html/form/upload/OnFilesSelectedBehavior.java:120]: Hum... Maybe I understood not correctly what @svenmeier was asking?

File: wicket-examples/src/main/java/org/apache/wicket/examples/ajax/builtin/EffectsPage.html
Comment (reiern70) [reply wicket-examples/src/main/java/org/apache/wicket/examples/ajax/builtin/EffectsPage.html:3]: Hm... no idea. Reverting back. Thanks for spotting this.

File: wicket-core/src/main/java/org/apache/wicket/markup/html/form/upload/OnFilesSelectedBehavior.java
Comment (martin-g) [reply wicket-core/src/main/java/org/apache/wicket/markup/html/form/upload/OnFilesSelectedBehavior.java:120]: @svenmeier said "file.add(OnFileSelectedBehavior.onSelected"
I am fine with either `onSelect` and `onSelected` but I like `onSelect` better, it is more consistent with the rest similar methods.

File: wicket-core/src/main/java/org/apache/wicket/markup/html/form/upload/OnFilesSelectedBehavior.java
Comment (reiern70) [reply wicket-core/src/main/java/org/apache/wicket/markup/html/form/upload/OnFilesSelectedBehavior.java:108]: No particular reason I think. I will fix.

File: wicket-core/src/main/java/org/apache/wicket/markup/html/form/upload/OnFilesSelectedBehavior.java
Comment (reiern70) [reply wicket-core/src/main/java/org/apache/wicket/markup/html/form/upload/OnFilesSelectedBehavior.java:46]: Thanks.

File: wicket-core/src/main/java/org/apache/wicket/markup/html/form/upload/OnFilesSelectedBehavior.java
Comment (reiern70) [reply wicket-core/src/main/java/org/apache/wicket/markup/html/form/upload/OnFilesSelectedBehavior.java:120]: image

?

File: wicket-core/src/main/java/org/apache/wicket/markup/html/form/upload/OnFilesSelectedBehavior.java
Comment (reiern70) [reply wicket-core/src/main/java/org/apache/wicket/markup/html/form/upload/OnFilesSelectedBehavior.java:108]: Fixed.

File: wicket-core/src/main/java/org/apache/wicket/markup/html/form/upload/OnFilesSelectedBehavior.java
Comment (reiern70) [reply wicket-core/src/main/java/org/apache/wicket/markup/html/form/upload/OnFilesSelectedBehavior.java:46]: fixed.

File: wicket-core/src/main/java/org/apache/wicket/markup/html/form/upload/OnFilesSelectedBehavior.java
Comment (martin-g) [reply wicket-core/src/main/java/org/apache/wicket/markup/html/form/upload/OnFilesSelectedBehavior.java:120]: Here, I think, he meant the class name, not the method name,

File: wicket-core/src/main/java/org/apache/wicket/markup/html/form/upload/OnFilesSelectedBehavior.java
Comment (reiern70) [reply wicket-core/src/main/java/org/apache/wicket/markup/html/form/upload/OnFilesSelectedBehavior.java:47]: done.

File: wicket-core/src/main/java/org/apache/wicket/markup/html/form/upload/OnFilesSelectedBehavior.js
Comment (reiern70) [reply wicket-core/src/main/java/org/apache/wicket/markup/html/form/upload/OnFilesSelectedBehavior.js:46]: Yes. I think we can.

File: wicket-examples/src/main/java/org/apache/wicket/examples/ajax/builtin/EffectsPage.html
Comment (reiern70) [reply wicket-examples/src/main/java/org/apache/wicket/examples/ajax/builtin/EffectsPage.html:3]: Fixed.

File: wicket-core/src/main/java/org/apache/wicket/markup/html/form/upload/OnFilesSelectedBehavior.js
Comment (reiern70) [reply wicket-core/src/main/java/org/apache/wicket/markup/html/form/upload/OnFilesSelectedBehavior.js:53]: Fixed.

File: wicket-core/src/main/java/org/apache/wicket/markup/html/form/upload/OnFilesSelectedBehavior.js
Comment (reiern70) [reply wicket-core/src/main/java/org/apache/wicket/markup/html/form/upload/OnFilesSelectedBehavior.js]: fixed

**Comment by reiern70** [comment]
@martin-g Many thanks for review and feedback. I have addressed your comments.

File: wicket-core/src/main/java/org/apache/wicket/markup/html/form/upload/OnFilesSelectedBehavior.js
Comment (reiern70) [reply wicket-core/src/main/java/org/apache/wicket/markup/html/form/upload/OnFilesSelectedBehavior.js]: Fixed this too

File: wicket-core/src/main/java/org/apache/wicket/markup/html/form/upload/OnFilesSelectedBehavior.java
Comment (reiern70) [reply wicket-core/src/main/java/org/apache/wicket/markup/html/form/upload/OnFilesSelectedBehavior.java]: renamed behavior to FilesSelectedBehavior and factory method to  onSelected.

File: wicket-core/src/main/java/org/apache/wicket/markup/html/form/upload/FileUploadField.java
Comment (reiern70) [reply wicket-core/src/main/java/org/apache/wicket/markup/html/form/upload/FileUploadField.java]: Fixed too.

**Comment by reiern70** [comment]
> > **Note:** A just thought that a possible addition in case of multiple file uploads is a component with a table displaying all the selected files. WDYT?
> 
> I don't think we should add such component. Every application may want to display it in a different way.

True. I have added some example on how to do that on ajax-upload page

File: wicket-core/src/main/java/org/apache/wicket/markup/html/form/upload/FilesSelectedBehavior.js
Comment (martin-g) [inline wicket-core/src/main/java/org/apache/wicket/markup/html/form/upload/FilesSelectedBehavior.js:44]: debug leftover

File: wicket-core/src/main/java/org/apache/wicket/markup/html/form/upload/FilesSelectedBehavior.js
Comment (martin-g) [inline wicket-core/src/main/java/org/apache/wicket/markup/html/form/upload/FilesSelectedBehavior.js:43]: you can move this line inside the `if` below

File: wicket-core/src/main/java/org/apache/wicket/markup/html/form/upload/FilesSelectedBehavior.java
Comment (martin-g) [inline wicket-core/src/main/java/org/apache/wicket/markup/html/form/upload/FilesSelectedBehavior.java:98]: For simplification you can use again JavaScript's `this` as in the precondition. And in the JS code : `inputField.id` to get the id.

**Review by martin-g — APPROVED** [review]

File: wicket-core/src/main/java/org/apache/wicket/markup/html/form/upload/FilesSelectedBehavior.java
Comment (reiern70) [reply wicket-core/src/main/java/org/apache/wicket/markup/html/form/upload/FilesSelectedBehavior.java:98]: Hum... I was not sure this would work: i.e. that in this case this will point to inputField, this is the reason I didn't try that. Giving it a try

File: wicket-core/src/main/java/org/apache/wicket/markup/html/form/upload/FilesSelectedBehavior.js
Comment (reiern70) [reply wicket-core/src/main/java/org/apache/wicket/markup/html/form/upload/FilesSelectedBehavior.js:44]: Do we have a Wicket.log? On our private project I have added a log function what can be enabled via some flag you can change at browser console (to switch on debugging).

File: wicket-core/src/main/java/org/apache/wicket/markup/html/form/upload/FilesSelectedBehavior.js
Comment (reiern70) [reply wicket-core/src/main/java/org/apache/wicket/markup/html/form/upload/FilesSelectedBehavior.js:43]: done.

File: wicket-core/src/main/java/org/apache/wicket/markup/html/form/upload/FilesSelectedBehavior.js
Comment (martin-g) [reply wicket-core/src/main/java/org/apache/wicket/markup/html/form/upload/FilesSelectedBehavior.js:44]: Yes - https://github.com/apache/wicket/blob/048e5681df960018722237ee6254273e839b5fd2/wicket-core/src/main/java/org/apache/wicket/ajax/res/js/wicket-ajax-jquery.js#L189-L222
It could be enabled/disabled via DebugSettings#ajaxDebugModeEnabled

File: wicket-core/src/main/java/org/apache/wicket/markup/html/form/upload/FilesSelectedBehavior.java
Comment (martin-g) [reply wicket-core/src/main/java/org/apache/wicket/markup/html/form/upload/FilesSelectedBehavior.java:98]: I might be wrong but I liked how you did it in `precondition()`!

Comment (reiern70) [reply wicket-core/src/main/java/org/apache/wicket/markup/html/form/upload/FilesSelectedBehavior.java:98]: This does not seems to refer to inputField in this case. I could use attrs.c as it is the ID but I do not see and extra benefit in this. Keeping as it is.

**Comment by reiern70** [comment]
@svenmeier and @martin-g Many thanks for reviews. Rebased and squashed commits. Merging now.

File: wicket-core/src/main/java/org/apache/wicket/markup/html/form/upload/FilesSelectedBehavior.js
Comment (reiern70) [reply wicket-core/src/main/java/org/apache/wicket/markup/html/form/upload/FilesSelectedBehavior.js]: I see... Thanks for info. I see you can maybe also type Wicket.Log.enable=true; in console to enable it for some page.

Comment (martin-g) [reply wicket-core/src/main/java/org/apache/wicket/markup/html/form/upload/FilesSelectedBehavior.js]: Yes! It will survive until the next page reload.

**Pull request merged**
